dvc/scm/git/backend/dulwich/__init__.py

```diff
@@ -342,6 +343,8 @@ def get_ref(self, name, follow: bool = True) -> Optional[str]:
             except ValueError:
                 pass
         if ref:
+            if isinstance(self.repo[ref], Tag):
+                ref = self.repo.get_peeled(name_b)
```

Comment:
This breaks the case where `name` is a symbolic reference and `follow = False`, like when `HEAD` points to `refs/heads/master`.  (see the current test failures)

Basically we need to catch `KeyError` when we do the `self.repo[ref]` check, and `pass` as long as `follow` is also `False`